#################### SETUP TAXSET ##########################

#########VARIABLES#############
#BEGIND = 10081 = The first I# folder within the TaxonSet folder; don't include the "I"
#ENDIND = 17900 = The last I# folder within the TaxonSet folder; don't include the "I"
#MINREADS = 20 = Minimum number of read coverage a locus must have to pass filter. Threshold found with plot generated by R identifyGoodSeqs
#ASSEMBLEDLOCI = 594 = Number of loci in probe design; found in Project's Assembly Methods
#TAXSET = T385 = "T#"
#NTHREADS = 15 = Number of threads/processors you want to use
#PLOIDYTAIL = C = determines whether to use PloidyspecificationFile_2 or PloidyspecificationFile_C
#MINSEQS = 11 = Minimum number of individuals a locus must have to pass filter, typically 1/2 the sample size
#NSCRIPTS = 15 = Number of cripts to run
#FOLDERTAIL = _conSeqs = foldertail of prealignments, alignments, trimmedalignments, and RAxML folders. Typically either "_conSeqs" or "_2alleles" depending on consensus sequences or phased alleles
#NEWNLOCI = 378 = New number of Loci _____; printed out after java SortSequencesViaOrthoSetsC. Locus conversion chart is found in prealignments folder
#NEWNTAXA = 23 = Number of samples. 2X the number of individuals/samples if Phased for 2 alleles.
#MISSINGALLOWED = 12 = Number of Missing sequences for alignments to tolerate. Typically 1/4 of NEWNTAXA 
#COLLABNAME = James_Clitelatta_leeches = Collaborator's name
#LINUX = LIN.UX = Last 3 digits of IP adress of machine that the Taxonset is stored on
#STORNUM = STOR.NUM = Storage number that the Taxonset is stored on
#DRIVENAME = DRIVE.NAME = Drive that the Taxonset will be offloaded to


###IF YOU DEVIATE FROM THIS TEMPLATE WRITE DOWN ANY CHANGES######

#This taxon set was done on LinuxIP LIN.UX






#YOU SHOULD BE IN THE TAXSET FOLDER ON THE STORAGE DRIVE
mkdir Results
mkdir Code
cd Code
cp '/home/alemmon/Dropbox/Anchor_Bioinformatics/Code/EvalInformativeSitesInPhylipMatrix.class' '/home/alemmon/Dropbox/Anchor_Bioinformatics/Code/GatherALLConSeqsWithOKCoverage2.class' '/home/alemmon/Dropbox/Anchor_Bioinformatics/Code/GetPairwiseDistanceMeasures.class' '/home/alemmon/Dropbox/Anchor_Bioinformatics/Code/IdentifyGoodSeqsViaReadsMappedC.r' '/home/alemmon/Dropbox/Anchor_Bioinformatics/Code/PlotAlignmentSummary2.r' '/home/alemmon/Dropbox/Anchor_Bioinformatics/Code/plotMDS5.r' '/home/alemmon/Dropbox/Anchor_Bioinformatics/Code/plotMDS5par.r' '/home/alemmon/Dropbox/Anchor_Bioinformatics/Code/SetupRAxML3.class' '/home/alemmon/Dropbox/Anchor_Bioinformatics/Code/SortSequencesViaOrthoSetsC.class' '/home/alemmon/Dropbox/Anchor_Bioinformatics/Code/toTaxonDrive.sh' '/home/alemmon/Dropbox/Anchor_Bioinformatics/Code/TrimAndMaskRawAlignments3.class' '/home/alemmon/Dropbox/Anchor_Bioinformatics/Code/writeOrthologyScript.sh' .

#IDENTIFY CONSENSUS SEQUENCES PASSING COVERAGE FILTER
R
source("IdentifyGoodSeqsViaReadsMappedC.r")
#identifyGoodSeqs(BEGIND,ENDIND,NHOMOLOGSTOCONSIDER,MINREADS)
#Threshold should be the 5th percentile (Blue line). Move Red line as close to Blue line as possible. 
identifyGoodSeqs(10081,17900,6,100)
quit()

#USE HOMOLOG IDS TO OBTAIN GOOD HOMOLOG SEQUENCES TO PUT IN HOMOLOGS FOLDER
#java -Xmx4000m GatherALLConSeqsWithOKCoverage2 ../TAXSET.txt ASSEMBLEDLOCI MINREADS
java -Xmx4000m GatherALLConSeqsWithOKCoverage2 ../T385.txt 594 20

#USE INFO IN HOMOLOGS FOLDER TO GENERATE DISTANCE MATRICIES
#java -Xmx4000m GetPairwiseDistanceMeasures TAXSET ASSEMBLEDLOCI KMERSIZE
java -Xmx4000m GetPairwiseDistanceMeasures T385 594 20



########################## ALLELE PHASING ##########################
#MAKE SURE THAT ALL INDIVIDUALS ARE ALLELE PHASED AT THIS POINT. 
#IF THIS PROJECT REQUIRES THEM TO BE ALLELE PHASED, CHANGE THE PLOIDY SPECIFICATION FILE TO _2 FROM _C 
rsync -av --progress '/home/alemmon/Dropbox/Anchor_Bioinformatics/Code/AllelePhaser3.class' '/home/alemmon/Dropbox/Anchor_Bioinformatics/Code/AllelePhaserRead.class' .
#IF MORE THAN 30, PARALLELIZE THE LOOP THIS WAY FOR 2 ALLELES
run-in-new-tab 'for i in {10081..17900}; do java AllelePhaser3 ../I$i/I$i 20000 0.5 2; done;'
#IF LESS THAN 30, PARALLELIZE THE LOOP THIS WAY FOR 2 ALLELES
for i in {10081..17900}; do run-in-new-tab "java AllelePhaser3 ../I$i/I$i 20000 0.5 2"; done;



########################## ORTHOLOGY ##########################
#USE DISTANCE MEASURES TO IDENTIFY ORTHOLOGOUS SETS
mkdir ../OrthoSets
chmod +x writeOrthologyScript.sh
./writeOrthologyScript.sh T385 594 15
chmod +x *Ortho*.sh
./allOrthoScripts.sh
./concatOrthology.sh

#SORT THE SEQUENCES INTO ORTHOLOGOUS SETS AND SETUP ALIGNMENTS
#MAKE SURE TO EDIT OR CHOOSE THE APPROPRIATE TAXSET PLOIDY SPECIFICATION FILE FOR THE CORRECT PHASED ALLELES
#java SortSequencesViaOrthoSetsC TAXSET SEQUENCESPECIFICATIONS MINSEQS NSCRIPTS FOLDERTAIL
java SortSequencesViaOrthoSetsC T385 ../T385_PloidySpecificationFile_C.txt 11 15 _conSeqs



########################## ALIGNMENT ##########################
#CHANGE PERMISSIONS
chmod +x *.sh
for i in {1..15}; do run-in-new-tab ./Align_AllViaMAFFT_conSeqs_${i}.sh; done;



########################## TRIMMING ##########################
#note the following block will be repeated until the following parameters are optimized: MINPROPSAME, MINGOODSITES, MISSINGALLOWED
#java TrimAndMaskRawAlignments3 TAXSET NEWNLOCI NEWNTAXA WINDOWSIZE MINGOODSITES MINPROPSAME MISSINGALLOWED ../TAXSET_LociAndTaxaToRemove.txt FOLDERTAIL
java TrimAndMaskRawAlignments3 T385 378 23 20 7 0.5 12 ../T385_LociAndTaxaToRemove.txt _conSeqs

#Rscript PlotAlignmentSummary2.r TAXSET NEWNTAXA NEWNLOCI MISSINGALLOWED FOLDERTAIL
Rscript PlotAlignmentSummary2.r "T385" 23 378 12 "_conSeqs"

#Look at plot Results/TAXSET_AlignmentSummary_nLoci.pdf adjust MISSINGALLOWED so it corrsponds to plateau
#Look at plot Results/TAXSET_AlignmentSummary_MaxTally.pdf adjust MINPROPSAME so it corrsponds to valley
#Look at plot Results/TAXSET_AlignmentSummary_WindowTally.pdf adjust MINGOODSITES so it corrsponds to valley

#java SetupRAxML3 TAXSET NEWNLOCI RAXMLJOBS RAXMLJOBS BOOTS BOOTS OUTGROUPTAXON1,OUTGROUPTAXON2 FOLDERTAIL
java SetupRAxML3 T385 378 6 6 100 100 _conSeqs

#Load RAxML/TAXSET_ConcatLoci.phylip in Geneious and inspect the alignment
#Increasing MINGOODSITES will increase masking.
#Increasing MINPROPSAME will increase trimming.

#Remove any bad loci or individuals in TAXSET_LociAndTaxaToRemove.txt
#Check RAxML_FOLDERTAIL/TAXSET_ConcatLoci.charSets to see which locus number needs to be removed
#Format:
LociNumber	Individual
348			ALL				#Removes locus 348 for all individuals.
ALL			21				#Removes all loci for individual 21. This means the 21st individual in Geneious.

#RUN ALL TRIMMING STEPS AGAIN IF BAD SEQUENCES IDENTIFIED



########################## PHYLOGENETICS ##########################
#SUMMARIZE FINAL ALIGNMENT
java EvalInformativeSitesInPhylipMatrix ../RAxML_conSeqs/T385_ConcatLoci.phylip > ../RAxML_conSeqs/T385_Statistics.txt && gedit ../RAxML_conSeqs/T385_Statistics.txt

#CHANGE DIRECTORIES AND PERMISSIONS
cd ../RAxML_conSeqs
chmod +x *.sh



############## CONCATENATED TREES ##############
./T385_ConcatLoci.sh



################ SEPARATE TREES ################
./T385_SeparateLoci.sh



#OPEN RAxML_bipartitions.TAXSET_ConcatLoci IN FIGTREE AND SAVE AS TAXSET_Tree.pdf



################# ASTRAL TREES #################
#MAKE ASTRAL FOLDER
mkdir ../Astral
cd ../Code
rsync -av --progress /home/alemmon/Dropbox/Anchor_Bioinformatics/Code/Astral4.9.7/lib /home/alemmon/Dropbox/Anchor_Bioinformatics/Code/Astral4.9.7/astral.4.9.7.jar .

#CONCATENATE EACH BESTTREE LOCI INTO ONE FILE
cat ../RAxML_conSeqs/RAxML_bestTree.T385_L* > ../Astral/RAxML_bestTree.T385_ALL.txt
#CONCATENATE EACH BOOTSTRAP LOCI INTO ONE FILE 
ls -d -1 ../RAxML_conSeqs/RAxML_bootstrap.T385_L* > ../Astral/bs-files

#RUN ASTRAL
java -jar astral.4.9.7.jar -i ../Astral/RAxML_bestTree.T385_ALL.txt -b ../Astral/bs-files -t 1 -o ../Astral/T385_Astral.tre

#OPEN IN FIGTREE AND MAKE PDF TREE
tail -n 1 ../Astral/T385_Astral.tre > ../Astral/AstralTree.tre



######################UPLOADING TO DROPBOX##########################
mkdir /home/alemmon/Dropbox/Anchor_Bioinformatics/TaxonSets/Results/T385_James_Clitelatta_leeches
mkdir /home/alemmon/Dropbox/Anchor_Bioinformatics/TaxonSets/Results/T385_James_Clitelatta_leeches/RAxML_conSeqs

#zip RAxMLFOLDERTAIL and upload files to TAXSET_COLLABNAME/RAxML DROPBOXFOLDER
sleep 0.1
run-in-new-tab cd /media/alemmon/storage*/T385
tar -zcvf RAxML_conSeqs.tar.gz RAxML_conSeqs
tar -zcvf Prealignments_conSeqs.tar.gz Prealignments_conSeqs
tar -zcvf Alignments_conSeqs.tar.gz Alignments_conSeqs
tar -zcvf TrimmedAlignments_conSeqs.tar.gz TrimmedAlignments_conSeqs

#rsyncs files to collab taxonset results folder
rsync -av --progress RAxML_conSeqs.tar.gz Prealignments_conSeqs.tar.gz Alignments_conSeqs.tar.gz TrimmedAlignments_conSeqs.tar.gz /home/alemmon/Dropbox/Anchor_Bioinformatics/TaxonSets/Results/T385_James_Clitelatta_leeches
rsync -av --progress T385_ConcatTree.pdf T385_ConcatTree_Cladogram.pdf ./RAxML_conSeqs/T385_ConcatLoci.charSets ./RAxML_conSeqs/T385_ConcatLoci.phylip ./RAxML_conSeqs/T385_Statistics.txt ./RAxML_conSeqs/RAxML_bipartitions.T385_ConcatLoci /home/alemmon/Dropbox/Anchor_Bioinformatics/TaxonSets/Results/T385_James_Clitelatta_leeches/RAxML_conSeqs
rsync -av --progress ./Astral /home/alemmon/Dropbox/Anchor_Bioinformatics/TaxonSets/Results/T385_James_Clitelatta_leeches



#MAKE SURE THE FOLLOWING FILES ARE ON DROPBOX

# FOR CONCATENATED TREES
# RAxML_conSeqs.zip
# TAXSET_ConcatLoci.charSets
# TAXSET_ConcatLoci.phylip
# TAXSET_Statistics.txt
#  RAxML_bipartitions.TAXSET_ConcatLoci 
#  TAXSET_Tree.pdf 


# FOR SEPARATE TREES
# RAxMLFOLDERTAIL.zip
# TAXSET_ConcatLoci.charSets
# TAXSET_ConcatLoci.phylip
# TAXSET_Statistics.txt


# FOR ASTRAL TREES
# Astral/bs-files
# Astral/RAxML_bestTree.TAXSET_ALL.txt
# Astral/TAXSET_Astral.tre
# Astral/AstralTree.tre
# Astral/AstralTree.pdf



##########################BACKUP DATA TAXONSET AND PROJECT DRIVES##########################
#Compress I numbers and move whole taxonset to Project Drive
#Upload to TaxonSets Drive
mkdir ../Samples
rsync -av --progress ../I* ../Samples
rm ../Samples/I*/*.fastq
tar -zcvf ../Samples.tar.gz ../Samples
rm -r ../Samples
rsync -avz --exclude 'I*' /media/alemmon/storage*/T385 alemmon@128.186.21.179:/Volumes/TaxonSets

#Upload to Collaborator ProjectDrive
rsync -avz --exclude 'I*' /media/alemmon/storage*/T385 alemmon@128.186.21.179:/Volumes/DRIVE.NAME








